### PR TITLE
Adding a deep copy implementation.

### PR DESCRIPTION
**Pull request by darynl**
This is useful to have so that you can preserve the original bid response but also manipulate a copy for logging purposes such as when ad markup is too large.

### CODE REVIEW

File: openrtb/bid.go [hunk 1]
```diff
@@ -43,3 +43,8 @@ func (bid Bid) Validate() error {
 
 	return nil
 }
+
+func (b *Bid) Copy() *Bid {
```
Comment (garukun) [inline openrtb/bid.go]: Why do we need pointer method receiver?

Comment (darynl) [reply openrtb/bid.go]: I guess we don't need it here since it's just a copy.

File: openrtb/bid.go [hunk 2]
```diff
@@ -43,3 +43,9 @@ func (bid Bid) Validate() error {
 
 	return nil
 }
+
+// Copy returns a pointer to a copy of the bid object.
+func (b Bid) Copy() *Bid {
+	bCopy := b
```
Comment (hwong) [inline openrtb/bid.go]: You still need to deep copy the slices (b.AdvertiserDomains, b.Categories, b.CreativeAttributes).

Comment (darynl) [reply openrtb/bid.go]: done

File: openrtb/bid_test.go [hunk 3]
```diff
@@ -1,9 +1,12 @@
 package openrtb_test
 
 import (
+	"encoding/json"
 	"reflect"
 	"testing"
 
+	"bytes"
```
Comment (hwong) [inline openrtb/bid_test.go]: nitpick: group+goimports

File: openrtb/bid.go [hunk 4]
```diff
@@ -43,3 +43,28 @@ func (bid Bid) Validate() error {
...
+
+	// Copy over slices
+	if b.AdvertiserDomains != nil {
+		bCopy.AdvertiserDomains = make([]string, len(b.AdvertiserDomains))
+		copy(bCopy.AdvertiserDomains, b.AdvertiserDomains)
+	}
```
Comment (hwong) [inline openrtb/bid.go]: nitpick: newlines after if blocks lol #stevepick

File: openrtb/bidresponse_test.go [hunk 5]
```diff
@@ -1,9 +1,12 @@
 package openrtb_test
 
 import (
+	"encoding/json"
 	"reflect"
 	"testing"
 
+	"bytes"
```
Comment (hwong) [inline openrtb/bidresponse_test.go]: nitpick: group for goimports

File: openrtb/bidresponse_test.go [hunk 6]
```diff
@@ -102,3 +105,58 @@ func TestBidResponseValidation(t *testing.T) {
...
+	for _, testCase := range testCases {
+		b2 := testCase.bidresponse.Copy()
+
+		if b2 == testCase.bidresponse {
+			t.Errorf("Address of bidresponse should not be the same in copied bidrequest. bidresponse1: %v bidresponse2: %v", &testCase.bidresponse, &b2)
+		}
+
+		for i := range testCase.bidresponse.SeatBids {
+			if &testCase.bidresponse.SeatBids[i] == &b2.SeatBids[i] {
+				t.Errorf("Address of seatbids should not be the same in copied bidrequest. seatbid1: %v seatbid2: %v.", &testCase.bidresponse.SeatBids[i], &b2.SeatBids[i])
```
Comment (hwong) [inline openrtb/bidresponse_test.go]: just fyi, you can use %p to print addresses, don't need to make a change here

File: openrtb/bid_test.go [hunk 7]
```diff
@@ -58,3 +61,86 @@ func TestBidValidation(t *testing.T) {
...
+		}
+
+		b2.ID = "1234"
+		if testCase.bid.ID == "1234" {
+			t.Errorf("Bid b2 should not be pointing to original bid object.\ntestCase: %+v\nCopy: %+v", testCase.bid, b2)
+		}
+
+		if len(b2.AdvertiserDomains) > 0 {
+			b2.AdvertiserDomains[0] = "456"
+			if testCase.bid.AdvertiserDomains[0] == "456" {
```
Comment (hwong) [inline openrtb/bid_test.go]: i like the address checking from below better, but up to you if you want to change it, no-op is fine

Comment (darynl) [reply openrtb/bid_test.go]: can I check the address? it's not a slice of pointers.

**Comment by darynl** [comment]
LGTM from @hwong

**Comment by darynl** [comment]
LGTM

**Pull request merged**